File: src/cli/commands/setup-global.js
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import { createRequire } from 'node:module';
import { fileURLToPath } from 'node:url';
import chalk from 'chalk';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const TEMPLATES_DIR = path.join(__dirname, '../../templates/commands');
const { version: PKG_VERSION } = createRequire(import.meta.url)('../../../package.json');

// Centralised platform registry — mirrors the pattern used by graphify.
// skill_dst is relative to os.homedir() unless it's an absolute path.
// version_stamp: write a .projctl_version file alongside the installed file.
const _PLATFORM_CONFIG = {
  claude: {
    template: 'projctl-claude.md',
    destDir: path.join(os.homedir(), '.claude', 'commands'),
    fileName: 'projctl.md',
    label: 'Claude Code',
  },
  cursor: {
    template: 'projctl-cursor.md',
    destDir: path.join(os.homedir(), '.cursor', 'commands'),
    fileName: 'projctl.md',
    label: 'Cursor',
  },
  windsurf: {
    template: 'projctl-windsurf.md',
    destDir: path.join(os.homedir(), '.codeium', 'windsurf', 'memories'),
    fileName: 'projctl.md',
    label: 'Windsurf',
  },
  copilot: {
    template: 'projctl-copilot.prompt.md',
    destDir: path.join(os.homedir(), '.copilot', 'prompts'),
    fileName: 'projctl.prompt.md',
    label: 'GitHub Copilot',
  },
};

function _versionFile(destDir) {
  return path.join(destDir, '.projctl_version');
}

function _checkInstalledVersion(destDir) {
  const vf = _versionFile(destDir);
  if (!fs.existsSync(vf)) return null;
  return fs.readFileSync(vf, 'utf8').trim();
}

function _writeVersionStamp(destDir) {
  fs.writeFileSync(_versionFile(destDir), PKG_VERSION, 'utf8');
}

function _refreshAllVersionStamps() {
  for (const cfg of Object.values(_PLATFORM_CONFIG)) {
    const vf = _versionFile(cfg.destDir);
    if (fs.existsSync(vf)) {
      try { fs.writeFileSync(vf, PKG_VERSION, 'utf8'); } catch { /* ignore */ }
    }
  }
}

export function checkInstalledVersions() {
  const results = [];
  for (const [key, cfg] of Object.entries(_PLATFORM_CONFIG)) {
    const dest = path.join(cfg.destDir, cfg.fileName);
    if (!fs.existsSync(dest)) continue;
    const installed = _checkInstalledVersion(cfg.destDir);
    results.push({ platform: key, label: cfg.label, dest, installed, current: PKG_VERSION });
  }
  return results;
}

export function setupGlobal(targets, opts = {}) {
  const results = [];

  for (const targetKey of targets) {
    const cfg = _PLATFORM_CONFIG[targetKey];
    if (!cfg) {
      results.push({ target: targetKey, status: 'unknown' });
      continue;
    }

    const src = path.join(TEMPLATES_DIR, cfg.template);
    if (!fs.existsSync(src)) {
      results.push({ target: targetKey, label: cfg.label, status: 'missing-template' });
      continue;
    }

    const dest = path.join(cfg.destDir, cfg.fileName);

    try {
      if (!opts.dryRun) {
        fs.mkdirSync(cfg.destDir, { recursive: true });
        fs.copyFileSync(src, dest);
        _writeVersionStamp(cfg.destDir);
      }
      results.push({ target: targetKey, label: cfg.label, dest, status: 'ok' });
    } catch (err) {
      results.push({ target: targetKey, label: cfg.label, status: 'error', error: err.message });
    }
  }

  if (!opts.dryRun && results.some(r => r.status === 'ok')) {
    _refreshAllVersionStamps();
  }

  return results;
}

export function registerSetupGlobalCommand(program) {
  const platformList = Object.keys(_PLATFORM_CONFIG).join(',');

  program
    .command('setup-global')
    .description('Install /projctl slash command globally for AI coding assistants')
    .option('--targets <targets>', `Comma-separated targets (${platformList})`, 'claude,cursor')
    .option('--dry-run', 'Preview without writing files')
    .option('--check', 'Show installed versions without installing')
    .action((opts) => {
      if (opts.check) {
        const versions = checkInstalledVersions();
        if (versions.length === 0) {
          console.log(chalk.dim('\n  /projctl is not installed in any known location.\n'));
          return;
        }
        console.log(chalk.bold('\n  projctl setup-global --check\n'));
        for (const v of versions) {
          const stale = v.installed && v.installed !== v.current;
          const badge = stale
            ? chalk.yellow(`v${v.installed} → upgrade to v${v.current}`)
            : chalk.green(`v${v.installed ?? 'unknown'} (up to date)`);
          console.log(`  ${chalk.bold(v.label.padEnd(18))} ${badge}`);
          console.log(`  ${chalk.dim(v.dest)}`);
        }
        console.log();
        return;
      }

      const targets = opts.targets.split(',').map(s => s.trim()).filter(Boolean);
      const unknown = targets.filter(t => !_PLATFORM_CONFIG[t]);
      if (unknown.length > 0) {
        console.error(chalk.red(`  Unknown target(s): ${unknown.join(', ')}`));
        console.error(chalk.dim(`  Valid targets: ${platformList}`));
        process.exit(1);
      }

      const results = setupGlobal(targets, { dryRun: opts.dryRun });

      console.log(chalk.bold('\n  projctl setup-global\n'));
      for (const r of results) {
        if (r.status === 'ok') {
          const icon = opts.dryRun ? chalk.dim('[dry-run]') : chalk.green('✓');
          console.log(`  ${icon} ${chalk.bold(r.label)}`);
          console.log(`     ${chalk.dim(r.dest)}`);
        } else if (r.status === 'error') {
          console.log(`  ${chalk.red('✗')} ${chalk.bold(r.label ?? r.target)}: ${r.error}`);
          console.error(chalk.dim(`     Tip: try running with elevated permissions or check write access to ${chalk.underline(path.dirname(r.dest ?? ''))}`));
        } else if (r.status === 'missing-template') {
          console.log(`  ${chalk.yellow('?')} ${chalk.bold(r.label ?? r.target)}: template not found — reinstall projctl`);
        } else {
          console.log(`  ${chalk.dim('?')} ${r.target}: unknown target`);
        }
      }

      const ok = results.filter(r => r.status === 'ok').length;
      if (ok > 0 && !opts.dryRun) {
        console.log(`\n  ${chalk.green(`/projctl is now available in ${ok} tool${ok !== 1 ? 's' : ''}`)}\n`);
        if (targets.includes('claude')) {
          console.log(chalk.dim('  Open Claude Code and type /projctl to get started.\n'));
        }
      }
    });
}

File: projhub/cli/setup_global.py
from __future__ import annotations
from importlib.metadata import version as _pkg_version
from pathlib import Path

import typer
from ._console import console

app = typer.Typer()

try:
    _VERSION = _pkg_version("projhub")
except Exception:
    _VERSION = "unknown"

_TEMPLATES_DIR = Path(__file__).parent.parent / "templates" / "commands"

_PLATFORM_CONFIG: dict[str, dict] = {
    "claude": {
        "template": "projhub-claude.md",
        "dest_dir": Path.home() / ".claude" / "commands",
        "file_name": "projhub.md",
        "label": "Claude Code",
    },
}


def _version_file(dest_dir: Path) -> Path:
    return dest_dir / ".projhub_version"


def _check_installed_version(dest_dir: Path) -> str | None:
    vf = _version_file(dest_dir)
    return vf.read_text(encoding="utf-8").strip() if vf.exists() else None


def _write_version_stamp(dest_dir: Path) -> None:
    _version_file(dest_dir).write_text(_VERSION, encoding="utf-8")


def _refresh_all_version_stamps() -> None:
    for cfg in _PLATFORM_CONFIG.values():
        vf = _version_file(cfg["dest_dir"])
        if vf.exists():
            try:
                vf.write_text(_VERSION, encoding="utf-8")
            except Exception:
                pass


def check_installed_versions() -> list[dict]:
    results = []
    for key, cfg in _PLATFORM_CONFIG.items():
        dest = cfg["dest_dir"] / cfg["file_name"]
        if not dest.exists():
            continue
        installed = _check_installed_version(cfg["dest_dir"])
        results.append({
            "platform": key,
            "label": cfg["label"],
            "dest": dest,
            "installed": installed,
            "current": _VERSION,
        })
    return results


def setup_global(targets: list[str], dry_run: bool = False) -> list[dict]:
    results = []
    for target in targets:
        cfg = _PLATFORM_CONFIG.get(target)
        if not cfg:
            results.append({"target": target, "status": "unknown"})
            continue

        src = _TEMPLATES_DIR / cfg["template"]
        if not src.exists():
            results.append({"target": target, "label": cfg["label"], "status": "missing-template"})
            continue

        dest = cfg["dest_dir"] / cfg["file_name"]
        try:
            if not dry_run:
                cfg["dest_dir"].mkdir(parents=True, exist_ok=True)
                import shutil
                shutil.copy(src, dest)
                _write_version_stamp(cfg["dest_dir"])
            results.append({"target": target, "label": cfg["label"], "dest": dest, "status": "ok"})
        except Exception as e:
            results.append({"target": target, "label": cfg["label"], "status": "error", "error": str(e)})

    if not dry_run and any(r["status"] == "ok" for r in results):
        _refresh_all_version_stamps()

    return results


@app.command("setup-global")
def setup_global_cmd(
    targets: str = typer.Option("claude", "--targets", help="Comma-separated targets (only 'claude' supported globally)"),
    dry_run: bool = typer.Option(False, "--dry-run", help="Preview without writing files"),
    check: bool = typer.Option(False, "--check", help="Show installed versions without installing"),
):
    """Install /projhub slash command globally for Claude Code.

    Note: Cursor, Windsurf, and Copilot don't support globally-installed slash
    commands. Use `projhub compile` per-project to generate their integration files.
    """
    if check:
        versions = check_installed_versions()
        if not versions:
            console.print("\n  [dim]/projhub is not installed in any known location.[/dim]\n")
            return
        console.print("\n  [bold]projhub setup-global --check[/bold]\n")
        for v in versions:
            stale = v["installed"] and v["installed"] != v["current"]
            badge = (
                f"[yellow]v{v['installed']} → upgrade to v{v['current']}[/yellow]"
                if stale
                else f"[green]v{v['installed'] or 'unknown'} (up to date)[/green]"
            )
            console.print(f"  [bold]{v['label']:<18}[/bold] {badge}")
            console.print(f"  [dim]{v['dest']}[/dim]")
        console.print()
        return

    target_list = [t.strip() for t in targets.split(",") if t.strip()]
    unknown = [t for t in target_list if t not in _PLATFORM_CONFIG]
    if unknown:
        valid = ", ".join(_PLATFORM_CONFIG)
        console.print(f"[red]  Unknown target(s): {', '.join(unknown)}[/red]")
        console.print(f"[dim]  Valid targets: {valid}[/dim]")
        raise typer.Exit(1)

    results = setup_global(target_list, dry_run=dry_run)

    console.print("\n  [bold]projhub setup-global[/bold]\n")
    for r in results:
        if r["status"] == "ok":
            icon = "[dim][dry-run][/dim]" if dry_run else "[green]✓[/green]"
            console.print(f"  {icon} [bold]{r['label']}[/bold]")
            console.print(f"     [dim]{r['dest']}[/dim]")
        elif r["status"] == "error":
            console.print(f"  [red]✗[/red] [bold]{r.get('label', r['target'])}[/bold]: {r['error']}")
        elif r["status"] == "missing-template":
            console.print(f"  [yellow]?[/yellow] [bold]{r.get('label', r['target'])}[/bold]: template not found — reinstall projhub")
        else:
            console.print(f"  [dim]?[/dim] {r['target']}: unknown target")

    ok_count = sum(1 for r in results if r["status"] == "ok")
    if ok_count > 0 and not dry_run:
        s = "s" if ok_count != 1 else ""
        console.print(f"\n  [green]/projhub is now available in {ok_count} tool{s}[/green]\n")
        if "claude" in target_list:
            console.print("[dim]  Open Claude Code and type /projhub to get started.[/dim]\n")
